src/transport/transport.cc

```diff
@@ -466,87 +480,156 @@ void ConvectionTransport::zero_time_step()
 }
 
 
-
-void ConvectionTransport::update_solution() {
-
-    START_TIMER("convection-one step");
-
-
-    START_TIMER("data reinit");
-    data_.set_time(time_->step()); // set to the last computed time
-
+bool ConvectionTransport::assess_time_constraint(double& time_constraint)
+{
     ASSERT(mh_dh, "Null MH object.\n" );
-    // update matrix and sources if neccessary
-
-
-    if (mh_dh->time_changed() > transport_matrix_time  || data_.porosity.changed()) {
+    data_.set_time(time_->step()); // set to the last computed time
+    
+    START_TIMER("data reinit");
+    
+    bool cfl_changed = false;
+    
+    // if FLOW or DATA changed ---------------------> recompute transport matrix
+    if (mh_dh->time_changed() > transport_matrix_time  || data_.porosity.changed())
+    {
         create_transport_matrix_mpi();
```

Comment:
This means that we do matrix assembly two times. 
What about splitting create_transport matrix_mpi into smaller functions (possibly inlined) for different tasks on single element. Then use only necessary parts to get CFL condition (matrix diagonal).